share/smack/top.py

```diff
@@ -275,204 +254,39 @@ def target_selection(args):
-    objc_flags = try_command(['gnustep-config', '--objc-flags'])
-    compile_command += objc_flags.split()
-  elif sys.platform == 'darwin':
-    sys.exit("Objective-C not yet supported on macOS")
-  else:
-    sys.exit("Objective-C not supported for this operating system.")
-  default_link_bc_files(compile_command, args)
-
-def d_frontend(args):
-  """Generate Boogie code from D programming language source(s)."""
-
-  # note: -g and -O0 are not used here. 
-  # Right now, it works, and with these options, smack crashes.
-  compile_command = ['ldc2', '-output-ll'] 
-  compile_command += map(lambda path: '-I=' + path, smack_headers())
-  args.entry_points = [ep if ep != 'main' else '_Dmain' for ep in args.entry_points]
-  default_link_bc_files(compile_command, args)
-
-def fortran_frontend(args):
-  """Generate Boogie code from Fortran language source(s)."""
-
-  # Requires two hacks as of right now:
-  #    1. The Debug Info Version in flang is incompatible with 
-  #       the version that clang uses. The workaround is to use
-  #       sed to change the file so llvm-link gives a warning
-  #       and not an error. This will be fixed in flang in the
-  #       near future (summer/fall 2018), after which this hack
-  #       will no longer be necessary.
-  #    2. For a fortran file that includes smack.f90 as a module,
-  #       it will not compile unless the file 'smack.mod' exists
-  #       in the working directory. 'smack.mod' is a build artifact
-  #       of compiling smack.f90. Therefore, the solution is to 
-  #       compile smack.f90 before the source files, even though
-  #       it is a smack library and could be in a generalized build_libs()
-
-  # replace the default entry point with the fortran default 'MAIN_'
-  args.entry_points = [ep if ep != 'main' else 'MAIN_' for ep in args.entry_points]
-
-  compile_command = default_clang_compile_command(args)
-  compile_command[0] = 'flang'
-  # compile to .ll instead of .bc in order to hack Debug Info Version
-  compile_command[1] = '-S'
+  libs = []
+  noreturning_frontend = False
   
-  bitcodes = []
+  def add_libs(lang):
+    if lang in extra_libs():
+      libs.append(extra_libs()[lang])
 
-  # must be done before compilation of input files
-  #   so that smack.mod appears
-  smack_fortran_lib = 'smack.f90'
-  smack_path = os.path.join(smack_lib(), smack_fortran_lib)
-  # use '.ll' instead of '.bc'
-  smack_fortran_bc = temporary_file(os.path.splitext(os.path.basename(smack_path))[0], '.ll', args)
-  try_command(compile_command + ['-o', smack_fortran_bc, smack_path], console=True)
-  # change the throw level of 'Debug Info Version' from error to warning in the IR
-  try_command(['sed', '-i', 's/i32 1, !\"Debug Info Version\"/i32 2, !\"Debug Info Version\"/g', smack_fortran_bc])
-  bitcodes.append(smack_fortran_bc)
+  if args.language:
+    lang = languages()[args.language]
+    if lang in ['BOOGIE', 'SVCOMP', 'JSON']:
+      noreturning_frontend = True
+
+    add_libs(lang)
```

Comment:
So, I don't think this problem is possible in the current implementation. If both C and C++ define and implement a function with the same name, then the whole program wouldn't compile normally anyways. If both C and C++ want to use __VERIFIER_nondet_int, or some other such symbol in smack.h, then they both import smack.h, and when compiled, each source file is compiled and will have a `declare void ....` external declaration in it's IR. Then, smack.c is compiled seperately, contains this function's implementation, and when all the files are linked everyone is happy. The key point is that the libraries are not necessarily the same language as the source. smack.c is never compiled as c++, and never should be. As another example, when I verify kotlin examples I need to use a C++ file and link it against the kotlin source in LLVM. 

There is a problem here, though, maybe this is what you were getting at, the data structure of `libs` is a Python list and not a set. Duplicate libraries could be added, which would break things. I will fix this promptly.

Reply:
I am not worried about the particulars of your Python code — though it’s a reasonable concern! — but about the following question: should we permit the case in which some function(s) have per-language implementations in cross-language projects?

I am not sure what the right answer is; my original thought was that each language can have its own header for a given function, but will ultimately link to the same implementation. But now I’m not sure that functions with the same name (e.g., `exit`) in different languages are supposed to have the same behavior.

Thoughts?

Reply:
Ok, I think I understand now. In my opinion, the difference is really between symbols that are declared by SMACK, and symbols that are language-specific but are modeled by SMACK. 

For language-specific symbols, I don't see how we can get around different implementations. For example, array and string implementations vary wildly between languages, and basically anything else in a standard library must be rewritten for every language (though maybe there is a way to double-dip for OBJC and swift, because they both use foundation?).

For SMACK-specifics like `assert`, `assume`, and the `nondet_*` family of functions, we are free to require the same implementation across languages, and I believe we should do so, because the basic functionality of verifier primitives will be the same across all languages. 

Maybe this is the problem we're having in #359. smack.h and smack.c should contain only smack-declared symbols. All other functions (e.g. exit, calloc) should be in seperate modeling files like `stdlib.c`, `string.c`, etc. When we generalize our `share/smack/lib/` folder, these modeling files can be demoted to only the C programming language. But smack.c contains the unique, cross-language implementations of basic verifier primitives. 

Of course, like you said, we can have language-specific headers without much trouble. My pet example is changing the header so that `__VERIFIER_assert` takes `BOOL` as it's parameter for objc, uses `bool` for C++, and has `_Bool` for everything else (e.g. with `_Bool` swift will bridge to its boolean type and not its Int type).

Reply:
Yes, I think I mostly agree with this point of view. Although I still wonder whether the ObjectiveC-Swift situation that you mention also surfaces, e.g., between C and C++.